### PR TITLE
Proxy metrics

### PR DESCRIPTION
**Pull request by lubennikovaav**
Implement the worker that periodically collects and sends proxy consumption metrics to the HTTP endpoint.
See spec in https://github.com/neondatabase/cloud/issues/3475

TODO (Need help)
- [x] Write a test. 
Locally running proxy used in python tests doesn't fill `endpoint_id` and other `traffic_labels`, so I don't know how to properly test it.. 
- [x] Set correct `node_id` for proxy. What should I use as a proxy's node_id?
- [x] fix metric collection thread cancellation
- [x] clarify API. Only send metrics for outbound traffic (tx). As discussed with @antonyc 
 ~~What should proxy report: separate metrics for each direction ( "rx"/"tx") or just one of them?~~

### CODE REVIEW

File: libs/consumption_metrics/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+///
+/// ```json
+/// {
+/// "metric": "remote_storage_size",
+/// "type": "absolute",
+/// "tenant_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+/// "timeline_id": "a03ebb4f5922a1c56ff7485cc8854143",
+/// "time": "2022-12-28T11:07:19.317310284Z",
+/// "idempotency_key": "2022-12-28 11:07:19.317310324 UTC-1-4019",
+/// "value": 12345454,
```
Comment (vadim2404) [inline libs/consumption_metrics/src/lib.rs:30]: `NIT`: trailing comma makes JSON invalid

Comment (lubennikovaav) [reply libs/consumption_metrics/src/lib.rs]: fixed

File: proxy/src/proxy_metrics.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        metric_collection_endpoint
+    );
+
+    // define client here to reuse it for all requests
+    let client = reqwest::Client::new();
+    let mut cached_metrics: HashMap<ProxyConsumptionMetricsKey, (u64, DateTime<Utc>)> =
+        HashMap::new();
+
+    loop {
+        tokio::select! {
```
Comment (lubennikovaav) [inline proxy/src/proxy_metrics.rs:39]: @funbringer 
could you please double check this code?

I'm not sure that I fully understand how `try_join_all` works.
Do I need to add some extra check for cancellation in this loop to ensure that it exits or it will happen automagically?

File: proxy/src/main.rs [hunk 3]
```diff
@@ -64,6 +65,18 @@ async fn main() -> anyhow::Result<()> {
...
+            Some(s) => Some(s.parse()?),
+            None => None,
+        };
+
+    let metric_collection_interval: std::time::Duration = humantime::parse_duration(
+        arg_matches
+            .get_one::<String>("metric-collection-interval")
+            .unwrap(),
+    )
+    .expect("cannot parse metric collection interval");
```
Comment (funbringer) [inline proxy/src/main.rs:79]: Please don't use `expect` like this. I think `?` is more appropriate here.

File: proxy/src/main.rs [hunk 4]
```diff
@@ -95,6 +108,8 @@ async fn main() -> anyhow::Result<()> {
     let config: &ProxyConfig = Box::leak(Box::new(ProxyConfig {
         tls_config,
         auth_backend,
+        metric_collection_endpoint,
+        metric_collection_interval,
```
Comment (funbringer) [inline proxy/src/main.rs:112]: There's a tls config and an auth backend config, which leads me to think that there should be metric collection config containing those two fields.

I'm not into prefixes.

File: proxy/src/main.rs [hunk 5]
```diff
@@ -126,6 +141,19 @@ async fn main() -> anyhow::Result<()> {
         )));
     }
 
+    if let Some(metric_collection_endpoint) = &config.metric_collection_endpoint {
+        let hostname = hostname::get().unwrap().into_string().unwrap();
```
Comment (funbringer) [inline proxy/src/main.rs:145]: Please don't use unwrap. This function returns anyhow::Result. Either use `.context(...)?` or just `?`.

File: proxy/src/main.rs [hunk 6]
```diff
@@ -126,6 +141,19 @@ async fn main() -> anyhow::Result<()> {
...
+
+        tasks.push(tokio::spawn(
+            proxy_metrics::collect_metrics(
+                metric_collection_endpoint,
+                config.metric_collection_interval,
```
Comment (funbringer) [inline proxy/src/main.rs:150]: I think we should extract this into `Option<MetricConfig>`. interval is useless without endpoint and vice versa.

File: test_runner/fixtures/neon_fixtures.py [hunk 7]
```diff
@@ -2301,6 +2301,20 @@ def extra_args(self) -> list[str]:
...
+            self.metric_collection_interval = metric_collection_interval
+
+        def extra_args(self) -> list[str]:
+            return [
+                # Link auth backend params
+                *["--auth-backend", "link"],
+                *["--uri", NeonProxy.link_auth_uri],
+                *["--metric-collection-endpoint", self.metric_collection_endpoint],
+                *["--metric-collection-interval", self.metric_collection_interval],
+            ]
```
Comment (funbringer) [inline test_runner/fixtures/neon_fixtures.py:2316]: This is misguided. There's no such thing as `link auth with metrics collector`. `AuthBackend` should've been a union of Postgres/Link/whatever, but there's no such notion in python, so we have to use inheritance. Those two params are universal and should be provided in init (default is None).

File: workspace_hack/Cargo.toml [hunk 8]
```diff
@@ -15,7 +15,7 @@ publish = false
 [dependencies]
 anyhow = { version = "1", features = ["backtrace", "std"] }
 bytes = { version = "1", features = ["serde", "std"] }
-chrono = { version = "0.4", default-features = false, features = ["clock", "iana-time-zone", "serde", "std", "winapi"] }
+chrono = { version = "0.4", features = ["clock", "iana-time-zone", "js-sys", "oldtime", "serde", "std", "time", "wasm-bindgen", "wasmbind", "winapi"] }
```
Comment (funbringer) [inline workspace_hack/Cargo.toml:18]: Why add new wasm-related features? This doesn't look right.

**Review by funbringer — COMMENTED** [review]
Thanks for the PR. I'll add more comments later, this is just a beginning.

File: workspace_hack/Cargo.toml (hunk 8 above)
Comment (lubennikovaav) [reply workspace_hack/Cargo.toml]: fixed
it was just unneeded chrono dependency

File: libs/consumption_metrics/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    /// This is an absolute, per-tenant metric.
+    RemoteStorageSize,
+    /// Logical size of the data in the timeline
+    /// This is an absolute, per-timeline metric
+    TimelineLogicalSize,
+    /// Number of bytes proxied
+    /// This is an incremental per endpoint proxy metric
+    /// see also NUM_BYTES_PROXIED_COUNTER
+    ProxyIOBytesPerClient,
+}
```
Comment (funbringer) [inline libs/consumption_metrics/src/lib.rs:157]: I still don't quite get why we need to define this in `consumption_metrics`. This library is a dependency and as such can't be responsible for enumerating all possible metrics of its users. Besides, that'd mean that e.g. `ProxyIOBytesPerClient` is available in pageserver, which doesn't make sense.

Instead, I propose do define metric names in a following manner:

```rust
// proxy/src/proxy/metrics.rs
const PROXY_IO_BYTES_PER_CLIENT: &str = "...";
...
```

Then it'd be possible to reuse those names everywhere in the proxy (the same logic applies elsewhere).

In `PageserverConsumptionMetric` and `ProxyConsumptionMetric` we'd have

```rust
// because the name should be a compile-time constant
pub metric: &'static str,
```

Furthermore, `PageserverConsumptionMetric` & `ProxyConsumptionMetric` look like twins. It should be possible to unite them and even extract the common denominator. Here's how I would do that:

```rust
use chrono::{DateTime, Utc};
use serde::Serialize;

fn main() {
    // This could be literally anything as long as it implements Serialize.
    // I'd put contextually-dependedent things here (e.g. things which make
    // sense for the proxy but not pageserver).
    //
    // I've noticed that in the cloud OpenAPI spec absolute events use
    // tenant_id while incremental events use endpoint_id, which to me
    // doesn't make sense. However, if that's how it should truly work,
    // we could move them to EventType.
    //
    // If there are any fields that I forgot to describe, we should
    // embed them into Event.
    #[derive(Serialize)]
    struct Ids {
        tenant_id: String,
        timeline_id: String,
    }

    let event = Event {
        kind: EventType::Incremental {
            start_time: Utc::now(),
            stop_time: Utc::now(),
        },
        metric: "metric name goes here",
        idempotency_key: "...".into(),
        value: 100500,
        extra: Ids {
            tenant_id: "foo".into(),
            timeline_id: "bar".into(),
        },
    };

    println!("{}", serde_json::to_string(&event).unwrap());
}

#[derive(Serialize, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
#[serde(tag = "type")]
pub enum EventType {
    #[serde(rename = "absolute")]
    Absolute {
        time: DateTime<Utc>,
    },
    #[serde(rename = "incremental")]
    Incremental {
        start_time: DateTime<Utc>,
        stop_time: DateTime<Utc>,
    },
}

#[derive(Serialize, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
pub struct Event<Extra> {
    #[serde(flatten)]
    pub kind: EventType,

    pub metric: &'static str,
    pub idempotency_key: String,
    pub value: u64,

    #[serde(flatten)]
    pub extra: Extra,
}
```

```
$ cargo run
   Compiling jsons v0.1.0 (/tmp/jsons)
    Finished dev [unoptimized + debuginfo] target(s) in 0.48s
     Running `target/debug/jsons`
{"type":"incremental","start_time":"2023-01-12T12:39:06.596256607Z","stop_time":"2023-01-12T12:39:06.596262568Z","metric":"metric name goes here","idempotency_key":"...","value":100500,"tenant_id":"foo","timeline_id":"bar"}
```

One of the benefits is that it closely mirrors the OpenAPI spec and is substantially easier to comprehend.

File: proxy/src/proxy_metrics.rs (hunk 2 above)
Comment (funbringer) [reply proxy/src/proxy_metrics.rs:39]: Not sure about `try_join_all`, it's not present here. Could you elaborate?

Anyways, currently there's no special cancellation logic in the proxy, so that shouldn't be a problem.

Speaking of `select!`, you shouldn't need here because you're polling just one future. So it should be:

```rust
    loop {
        let _ = ticker.tick().await;

        let res = collect_metrics_task(
            &client,
            &mut cached_metrics,
            metric_collection_endpoint,
            hostname.clone(),
        )
        .await;

        match res {
            Err(e) => {
                error!("Failed to send consumption metrics: {} ", e);
            }
            Ok(_) => {
                trace!("collect_metrics_task completed successfully")
            }
        }
    }
```

However, let's take a closer look at the error handling part. Do you think we should terminate proxy if we are unable to collect or send metrics? I don't think that's a good idea, because clients are unlikely to be happy about that. We should consume the error and log it.

File: proxy/src/proxy_metrics.rs [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                    (value, Utc::now()),
+                ));
+            }
+        }
+    }
+
+    current_metrics
+}
+
+pub async fn collect_metrics_task(
```
Comment (funbringer) [inline proxy/src/proxy_metrics.rs:96]: naming nit: It's not a task, because "task" means the top-level future (aka green thread).

File: proxy/src/proxy_metrics.rs [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    client: &reqwest::Client,
+    cached_metrics: &mut HashMap<ProxyConsumptionMetricsKey, (u64, DateTime<Utc>)>,
+    metric_collection_endpoint: &reqwest::Url,
+    hostname: String,
```
Comment (funbringer) [inline proxy/src/proxy_metrics.rs:100]: Why String instead of &str though?

File: libs/consumption_metrics/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ProxyConsumptionMetricsKey {
+    pub endpoint_id: String,
+    pub direction: String,
+}
+
+#[derive(serde::Serialize)]
+pub struct ProxyEventChunk<'a> {
+    pub events: &'a [ProxyConsumptionMetric],
+}
```
Comment (funbringer) [inline libs/consumption_metrics/src/lib.rs:221]: I don't think we need this. We wouldn't be able to derive Serialize if there was no way to serialize the slice itself.

File: proxy/src/proxy_metrics.rs [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+                    .find(|l| l.get_name() == "direction")
+                    .unwrap()
+                    .get_value();
+
+                let endpoint_id = ms
+                    .get_label()
+                    .iter()
+                    .find(|l| l.get_name() == "endpoint_id")
+                    .unwrap()
+                    .get_value();
```
Comment (funbringer) [inline proxy/src/proxy_metrics.rs:74]: It's possible to do this in one traversal using `for` and two options.

File: libs/consumption_metrics/src/lib.rs (hunk 12 above)
Comment (lubennikovaav) [reply libs/consumption_metrics/src/lib.rs:221]: This is a way to serialize chunks into jsons of the following format
`{"events" : [ {***}, {***} ] }`
instead of manual json constuction.

Comment (funbringer) [reply libs/consumption_metrics/src/lib.rs:221]: Ah, right, I forgot about this. Still, this doesn't have to be tightly coupled with proxy's metrics:

```rust
#[derive(serde::Serialize)]
struct Events<'a, T> {
    pub events: &'a [T],
}
```

File: proxy/src/proxy_metrics.rs (hunk 2 above)
Comment (lubennikovaav) [reply proxy/src/proxy_metrics.rs:39]: sorry for the confusing question.
I was referring to [this code](https://github.com/neondatabase/neon/blob/9d6f88360eb4703ac42e293276307b083fd73e99/proxy/src/main.rs#L167) in in proxy main.rs
```
   // This will block until all tasks have completed.
    // Furthermore, the first one to fail will cancel the rest.
    let _: Vec<()> = futures::future::try_join_all(tasks).await?;
```

but `collect_metrics` future just loops forever and doesn't check for any shutdown signal.
Will it be somehow cancelled by tokio when other futures from the set fail, or I need to implement shutdown logic in `collect_metrics` so that main proxy thread would not hang forever?

File: proxy/src/proxy_metrics.rs (hunk 10 above)
Comment (lubennikovaav) [reply proxy/src/proxy_metrics.rs]: renamed to `collect_metrics_iteration`

File: libs/consumption_metrics/src/lib.rs (hunk 9 above)
Comment (lubennikovaav) [reply libs/consumption_metrics/src/lib.rs]: Thank you for the suggestion!
I pushed the refactoring

File: proxy/src/proxy_metrics.rs [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+                            // cache new metric
+                            .or_insert((send_metric.value, send_metric.stop_time));
+                    }
+                } else {
+                    error!("metrics endpoint refused the sent metrics: {:?}", res);
+                }
+            }
+            Err(err) => {
+                error!("failed to send metrics: {:?}", err);
+            }
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply proxy/src/proxy_metrics.rs]: Sure. We've discussed it before and agreed that retries can be implemented later.
We send metrics often enough and cache only what was sent successfully, so small glitches shouldn't be a big deal.

File: proxy/src/proxy_metrics.rs (hunk 13 above)
Comment (lubennikovaav) [reply proxy/src/proxy_metrics.rs]: there are not too many labels, so I'd keep it as is.
to me that looks more readable than alternative

File: libs/consumption_metrics/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+//!
+
+use utils::id::TimelineId;
+
+use utils::id::TenantId;
+
+use serde::Serialize;
+
+use chrono::{DateTime, Utc};
+use rand::Rng;
```
Comment (funbringer) [inline libs/consumption_metrics/src/lib.rs:12]: nit: needless spaces; worth checking other files too.

File: libs/consumption_metrics/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+///
+/// Both the proxy and the ingestion endpoint will live in the same region (or cell)
+/// so while the project-id is unique across regions the whole pipeline will work correctly
+/// because we enrich the event with project_id in the control-plane endpoint.
+///
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ProxyConsumptionMetricsKey {
+    pub endpoint_id: String,
+    pub direction: String,
+}
```
Comment (funbringer) [inline libs/consumption_metrics/src/lib.rs:110]: We could do this:

```rust
pub enum ProxyConsumptionMetricsKey {
  Tx { endpoint_id: String },
  Rx { endpoint_id: String },
}
```

File: libs/consumption_metrics/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    )
+}
+
+/// Key that uniquely identifies the object, this metric describes.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PageserverConsumptionMetricsKey {
+    pub tenant_id: TenantId,
+    pub timeline_id: Option<TimelineId>,
+    pub metric: &'static str,
+}
```
Comment (funbringer) [inline libs/consumption_metrics/src/lib.rs:88]: What do you think about moving this to the pageserver? Same for `ProxyConsumptionMetricsKey`. That'd make this crate even smaller, but I'm reasonably confident that there is some logic that we could put here (maybe later).

File: proxy/src/metrics.rs [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+                            // cache new metric
+                            .or_insert((send_metric.value, stop_time));
+                    }
+                } else {
+                    error!("metrics endpoint refused the sent metrics: {:?}", res);
+                }
+            }
+            Err(err) => {
+                error!("failed to send metrics: {:?}", err);
+            }
```
Comment (funbringer) [inline proxy/src/metrics.rs:195]: It's possible to decrease nesting level here:

```rust
for ... {
    let Ok(res) = res else {
        error!("failed to send metrics: {:?}", err);
        continue;
    };
    
    // yay, res is available here!
}
```

**Review by funbringer — COMMENTED** [review]
Thanks! We're almost there :+1:

File: proxy/src/metrics.rs (hunk 18 above)
Comment (lubennikovaav) [reply proxy/src/metrics.rs:195]: nice!
though, how do I get the `err` using this syntax?

File: libs/consumption_metrics/src/lib.rs (hunk 16 above)
Comment (lubennikovaav) [reply libs/consumption_metrics/src/lib.rs:110]: I removed this struct at all, as it is replaced by `Ids`.

File: proxy/src/metrics.rs (hunk 18 above)
Comment (funbringer) [reply proxy/src/metrics.rs:195]: Ah, right... I'm getting tired.

```rust
let res = match res {
  Ok(x) => x,
  Err(e) => { error!(...); continue; }
}
```

**Review by funbringer — APPROVED** [review]
LGTM, thank you! Now we just need to rebase.

File: pageserver/src/consumption_metrics.rs [hunk 19]
```diff
@@ -300,35 +200,29 @@ pub async fn collect_metrics_task(
 
     // Send metrics.
     // Split into chunks of 1000 metrics to avoid exceeding the max request size
-    const CHUNK_SIZE: usize = 1000;
     let chunks = current_metrics.chunks(CHUNK_SIZE);
 
-    let mut chunk_to_send: Vec<ConsumptionMetric> = Vec::with_capacity(1000);
+    let mut chunk_to_send: Vec<Event<Ids>> = Vec::with_capacity(1000);
```
Comment (problame) [inline pageserver/src/consumption_metrics.rs:205]: `1000` magic constant. Is this `CHUNK_SIZE`?

Also, should update the `1000` in the comment above.

File: pageserver/src/consumption_metrics.rs [hunk 20]
```diff
@@ -300,35 +200,29 @@ pub async fn collect_metrics_task(
 
     // Send metrics.
```
Comment (problame) [inline pageserver/src/consumption_metrics.rs:201]: Why can't we share everything from here on with `proxy`? Looks quite similar to me.

**Review by problame — COMMENTED** [review]
Only looked at pageserver changes.

File: pageserver/src/consumption_metrics.rs (hunk 20 above)
Comment (lubennikovaav) [reply pageserver/src/consumption_metrics.rs:201]: umm, we collect metrics in a different way and they have slightly different format.

all shared code is already moved in lib/consumption_metrics

File: pageserver/src/consumption_metrics.rs (hunk 19 above)
Comment (lubennikovaav) [reply pageserver/src/consumption_metrics.rs]: thank you! fixed

File: pageserver/src/consumption_metrics.rs [hunk 21]
```diff
@@ -300,35 +200,29 @@ pub async fn collect_metrics_task(
...
     // Split into chunks of 1000 metrics to avoid exceeding the max request size
-    const CHUNK_SIZE: usize = 1000;
     let chunks = current_metrics.chunks(CHUNK_SIZE);
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply pageserver/src/consumption_metrics.rs:203]: i'd leave it as is. 
qualified name looks a bit too wordy to me

File: pageserver/src/consumption_metrics.rs (hunk 20 above)
Comment (problame) [reply pageserver/src/consumption_metrics.rs:201]: I'm speaking of just the chunking + sending part. Those look the same to my eyes.

I won't die on that hill, though.

File: pageserver/src/consumption_metrics.rs (hunk 19 above)
Comment (problame) [reply pageserver/src/consumption_metrics.rs]: Comment  is still missing:

```
// Split into chunks of 1000 metrics to avoid exceeding the max request size
```

**Review by problame — APPROVED** [review]

File: pageserver/src/consumption_metrics.rs (hunk 20 above)
Comment (funbringer) [reply pageserver/src/consumption_metrics.rs:201]: I like @problame's proposal though. Can we leave a TODO comment? I'm sure somebody will pick this up later.

**Review by funbringer — APPROVED** [review]
You probably need to update helm charts to cover new cmdline args. However, this can be done a separate PR.
